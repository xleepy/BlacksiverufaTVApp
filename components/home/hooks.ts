import { useEffect, useState } from 'react';
import { SEGMENTS_PATH } from '@env';

export type Segment = {
  youtube?: string;
  name: string;
  date: string;
  games: string[];
  note?: string;
  direct?: string;
  hls?: string;
  subtitles?: string;
  id: string;
};

export function useSegments(): [Segment[], boolean] {
  const [segments, setSegments] = useState<Segment[]>([]);
  const [isLoading, setLoading] = useState(false);
  useEffect(() => {
    let ignore = false;
    setLoading(true);
    fetch(SEGMENTS_PATH)
      .then(res => res.json())
      .then(result => {
        if (ignore) {
          return;
        }
        const allVideos: Segment[] = Object.entries<Record<string, Segment>>(
          result,
        ).map(([key, segment]) => {
          return { ...segment, id: key } as Segment;
        });
        setSegments(allVideos);
      })
      .finally(() => {
        setLoading(false);
      });

    return () => {
      ignore = true;
    };
  }, []);

  return [segments, isLoading];
}
