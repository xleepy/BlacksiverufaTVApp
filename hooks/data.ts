import { useEffect, useState } from 'react';
import { SEGMENTS_PATH } from '@env';

export type Segment = {
  youtube: string;
  name: string;
  date: string;
  games: string[];
  note?: string;
};

export function useSegments() {
  const [segments, setSegments] = useState<Segment[]>();
  useEffect(() => {
    fetch(SEGMENTS_PATH)
      .then(res => res.json())
      .then(result => {
        const allVideos: Segment[] = Object.values(result);
        setSegments(allVideos);
      });
  }, []);

  return segments ?? [];
}
