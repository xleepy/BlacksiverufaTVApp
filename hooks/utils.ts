import { useCallback, useEffect, useMemo, useRef } from 'react';
import { useWindowDimensions } from 'react-native';
type Orientation = 'portrait' | 'landscape';

export function useOrientation(): Orientation {
  const windowDimensions = useWindowDimensions();
  return useMemo<Orientation>(() => {
    return windowDimensions.height >= windowDimensions.width
      ? 'portrait'
      : 'landscape';
  }, [windowDimensions]);
}

export function useEventCallback(
  fn: (...args: any) => any,
  dependencies: any[],
) {
  const ref = useRef<(...args: any[]) => any>(() => {
    throw new Error('Cannot call an event handler while rendering.');
  });

  useEffect(() => {
    ref.current = fn;
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [fn, ...dependencies]);

  return useCallback(() => {
    const currentFn = ref.current;
    return currentFn();
  }, []);
}
