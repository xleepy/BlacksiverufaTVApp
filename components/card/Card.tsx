import React, { useCallback, useMemo, memo } from 'react';
import { Segment } from '../home/hooks';
import {
  ImageBackground,
  ImageProps,
  StyleSheet,
  Text,
  TouchableHighlight,
  View,
} from 'react-native';

type Props = {
  segment: Segment;
  onSegmentSelect: (segment: Segment) => void;
};

const styles = StyleSheet.create({
  card: {
    borderBottomColor: 'black',
    padding: 10,
    gap: 10,
    height: 220,
    borderColor: 'rgba(0,0,0,.125)',
    borderWidth: 1,
    borderStyle: 'solid',
    borderRadius: 0.25,
  },
  image: {
    width: '100%',
    height: 200,
  },
  text: {
    backgroundColor: 'rgb(52, 58, 64)',
    color: 'white',
    marginTop: 'auto',
    padding: 8,
    textAlign: 'justify',
  },
});

export const Card = memo(({ segment, onSegmentSelect }: Props) => {
  const handlePress = useCallback(() => {
    onSegmentSelect(segment);
  }, [onSegmentSelect, segment]);

  const source = useMemo<ImageProps['source']>(() => {
    if (segment.youtube) {
      return {
        uri: `https://img.youtube.com/vi/${segment.youtube}/mqdefault.jpg`,
        cache: 'default',
      };
    }
    return {
      uri: `https://bsufiles.drhx.ru/streams/${segment.id}-sm.jpg`,
      cache: 'default',
    };
  }, [segment]);

  return (
    <TouchableHighlight onPress={handlePress}>
      <View testID={segment.name} style={styles.card}>
        <ImageBackground
          resizeMode="cover"
          style={styles.image}
          source={source}>
          <Text style={styles.text}>{segment.name}</Text>
        </ImageBackground>
      </View>
    </TouchableHighlight>
  );
});
