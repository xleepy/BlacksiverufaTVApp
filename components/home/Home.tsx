import React, { useCallback, useEffect } from 'react';
import { Segment, useSegments } from './hooks';
import {
  ListRenderItem,
  SafeAreaView,
  StyleSheet,
  Text,
  VirtualizedList,
} from 'react-native';
import { Card } from '../card/Card';
import { NativeStackScreenProps } from '@react-navigation/native-stack';

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
});

export const Home = ({ navigation }: NativeStackScreenProps<any, any>) => {
  const [segments, isLoading] = useSegments();

  useEffect(() => {
    navigation.setOptions({ title: 'Архив стримов BlackUFA' });
  }, [navigation]);

  const handleNavigationToPlayer = useCallback(
    (segment: Segment) => {
      navigation.navigate('Player', segment);
    },
    [navigation],
  );

  const renderItem = useCallback<ListRenderItem<Segment>>(
    ({ item }) => {
      return (
        <Card
          key={item.id}
          segment={item}
          onSegmentSelect={handleNavigationToPlayer}
        />
      );
    },
    [handleNavigationToPlayer],
  );

  const getItemCount = useCallback(() => {
    return segments.length;
  }, [segments]);

  const getItem = useCallback((data: Segment[], idx: number) => {
    return data[idx];
  }, []);

  const keyExtractor = useCallback((item: Segment) => {
    return item.id;
  }, []);

  return (
    <SafeAreaView testID="home-view" style={styles.container}>
      {isLoading && <Text>Loading...</Text>}
      {segments.length > 0 && (
        <VirtualizedList
          testID="segments-list"
          removeClippedSubviews={true}
          data={segments}
          initialNumToRender={3}
          maxToRenderPerBatch={3}
          updateCellsBatchingPeriod={300}
          keyExtractor={keyExtractor}
          getItem={getItem}
          getItemCount={getItemCount}
          renderItem={renderItem}
        />
      )}
    </SafeAreaView>
  );
};
