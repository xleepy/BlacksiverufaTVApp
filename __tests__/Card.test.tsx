import 'react-native';
import React from 'react';
import { render } from '@testing-library/react-native';
import { Card } from '../components/card/Card';
import { Segment } from '../components/home/hooks';

describe('Card component tests', () => {
  const mockSegment: Segment = {
    id: 'test',
    games: [],
    date: '2015-08-10',
    name: 'test',
  };
  const pressMock = jest.fn();

  it('should render', () => {
    const { toJSON } = render(
      <Card segment={mockSegment} onSegmentSelect={pressMock} />,
    );
    expect(toJSON()).toMatchSnapshot();
  });
});
