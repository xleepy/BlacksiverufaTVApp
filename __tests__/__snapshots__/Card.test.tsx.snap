// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card component tests should render 1`] = `
<View
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
>
  <View
    style={
      {
        "borderBottomColor": "black",
        "borderColor": "rgba(0,0,0,.125)",
        "borderRadius": 0.25,
        "borderStyle": "solid",
        "borderWidth": 1,
        "gap": 10,
        "height": 220,
        "padding": 10,
      }
    }
    testID="test"
  >
    <View
      accessibilityIgnoresInvertColors={true}
      style={
        {
          "height": 200,
          "width": "100%",
        }
      }
    >
      <Image
        resizeMode="cover"
        source={
          {
            "cache": "default",
            "uri": "https://bsufiles.drhx.ru/streams/test-sm.jpg",
          }
        }
        style={
          [
            {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            },
            {
              "height": 200,
              "width": "100%",
            },
            undefined,
          ]
        }
      />
      <Text
        style={
          {
            "backgroundColor": "rgb(52, 58, 64)",
            "color": "white",
            "marginTop": "auto",
            "padding": 8,
            "textAlign": "justify",
          }
        }
      >
        test
      </Text>
    </View>
  </View>
</View>
`;
